### ISSUE TITLE
Why ankle torque minimization in wrench distribution?

### ISSUE THREAD

**Issue by shbang91** [jrl-umi3218/lipm_walking_controller#44]
Hello,

Thank you for your effort in maintaining this nice repo!

I have a question regarding the ankle torque minimization (Eq.(13)) in II-C. Contact Wrench Distribution on your stair climbing stabilization paper. 

First of all, what is the benefit of minimizing ankle torque minimization (Eq.(13)) over minimizing the contact wrench in each sole center frame (lc_W_left, rc_W_right)? (I believe the position of the sole center frame(lc, rc) is in the bottom center of the foot which directly contacts with the ground. Please correct me if I am wrong).

Secondly,  looking at the https://github.com/stephane-caron/lipm_walking_controller/blob/34f5d1e03d49ca99887d326de25c8a4357f2c451/src/Stabilizer.cpp#L617, I wonder why you weighted more on the taux and tauy. Is this because you don't want to generate tilting motions in the x and y local ankle frame (By minimizing the external wrench applied on the ankle joint through Eq.(13), you would expect the ankle joint to execute small torque command)?

Finally, do the sole center frame and ankle frame share the same orientation in your definition?

Thank you,
Seung Hyeon

**Comment by stephane-caron** [comment]
> First of all, what is the benefit of minimizing ankle torque minimization (Eq.(13)) over minimizing the contact wrench in each sole center frame (lc_W_left, rc_W_right)?

Ankle torque minimization is motivated by the design of the soles of HRP robots, which is inherited from Honda's humanoid robot design:

![image](https://user-images.githubusercontent.com/1189580/198240757-361a48e0-3d9d-46f2-a4c7-622d57ccd531.png)

(Source: slides 4 and 5 in [this presentation](https://scaron.info/slides/jpl-2019.pdf#page=5).)

The mechanical flexibility below the ankles is here to reduce the rigidity of the contact interaction with the environment. This means protecting the force sensor from hard impacts, but also importantly reducing the control frequency required for foot admittance control.

One downside of this flexibility is that our sensing and admittance control laws are based on the assumption that *its deflection is small*. (To be precise, the deflection is the orientation of the sole frame with respect to the ankle frame.) The larger the deflection, the worse our control becomes. That's why most works dealing with HRP robots aim to keep this deflection small.

That's also the motivation for projects like [SpringEstimator](https://github.com/jrl-umi3218/SpringEstimator) which observe the deflection directly. In this controller, we assume that there exists [a linear underlying model](https://scaron.info/publications/icra-2019.html#comment-reply-damping-control-at-foot-contacts) $$\Delta_{spring} = \frac{1}{K} \tau_{ankle},$$ and minimize the ankle torque $\tau_{ankle}$ as a proxy to minimize the deflection $\Delta_{spring}$.

**Comment by stephane-caron** [comment]
> Secondly, looking at the [lipm_walking_controller/src/Stabilizer.cpp](https://github.com/stephane-caron/lipm_walking_controller/blob/34f5d1e03d49ca99887d326de25c8a4357f2c451/src/Stabilizer.cpp#L617), I wonder why you weighted more on the taux and tauy. Is this because you don't want to generate tilting motions in the x and y local ankle frame (By minimizing the external wrench applied on the ankle joint through Eq.(13), you would expect the ankle joint to execute small torque command)?

The coordinates $\tau_{ankle,x/y}$, in full: "the moment of ground contact forces around the local $x$ and $y$ axes of the (projected) ankle frame", correspond to the center of pressure via the [cross-product formula](https://scaron.info/robot-locomotion/zero-tilting-moment-point.html#center-of-pressure):

$$
p_{CoP} = p_{ankle} + \frac{\tau_{ankle} \times e_z}{f \cdot e_z},
$$

where $e_z$ is the normal vector of the projected ankle frame (orthogonal to the sole surface). The motivation for weighing those two coordinates more than the third one $(\tau_{ankle,z})$ is to give priority to center-of-pressure control over yaw-moment control. There are at least two reasons why CoP control is more important:

1. Centers of pressure under each contact contribute directly to the net ZMP, which is what we aim to regulate in the end.
2. Centers of pressure contribute indirectly to the [yaw-moment friction condition](https://scaron.info/pages/ideas.html#yaw-friction-constraint-for-planar-contacts):

$$
|\tau_z - \tau_z^*| \leq \mu f_z d_{CoP},
$$ 

where $d_{CoP}$ is the distance between the CoP and the edge of the contact surface, which we try to keep as large as possible (by minimizing ankle torques, *i.e.* keeping each CoP below its corresponding ankle). The larger $d_{CoP}$ is, the less the robot has to pay attention to $\tau_{ankle,z}$ as yaw slippage becomes less likely.

**Comment by stephane-caron** [comment]
> Finally, do the sole center frame and ankle frame share the same orientation in your definition?

Yes, the sole center frame and projected ankle frame of a foot share the same orientation.

**Comment by stephane-caron** [comment]
Hi @shbang91, I've posted some answers to your questions at https://github.com/stephane-caron/lipm_walking_controller/discussions/77. Hoping this helps!
